File: hackerrank/problem solving/algorithms/2 - implementation/22 - angry professor/angry_professor.c
// https://www.hackerrank.com/challenges/angry-professor/problem?isFullScreen=true
// C23

#include <stdio.h>

void read_numbers(int *arr, int n);
bool angry_professor(const int *arrival_times, int n, int threshold);

int main()
{
    int n;
    scanf("%d", &n);

    for (int i = 0; i < n; ++i) {
        int size, threshold;
        scanf("%d %d", &size, &threshold);
        int arrival_times[size];
        read_numbers(arrival_times, size);
        puts(angry_professor(arrival_times, size, threshold) ? "YES" : "NO");
    }

    return 0;
}

void read_numbers(int *arr, int n)
{
    for (int i = 0; i < n; ++i)
        scanf("%d", &arr[i]);
}

// k: length of arrival_times
// T: O(k)
// S: O(1) extra space
bool angry_professor(const int *arrival_times, int n, int threshold)
{
    int on_time = 0;
    for (int i = 0; i < n; ++i)
            on_time += arrival_times[i] <= 0;
    return on_time < threshold;
}

File: hackerrank/problem solving/algorithms/2 - implementation/22 - angry professor/angry_professor.go
// https://www.hackerrank.com/challenges/angry-professor/problem?isFullScreen=true

package main

import "fmt"

func main() {
    var n int
    _, _ = fmt.Scan(&n)

    for range n {
        var size, threshold int
        _, _ = fmt.Scan(&size, &threshold)
        arrivalTimes := readNumbers(size)

        if angryProfessor(arrivalTimes, threshold) {
            fmt.Println("YES")
        } else {
            fmt.Println("NO")
        }
    }
}

func readNumbers(n int) []int {
    arr := make([]int, n)
    for i := range n {
        _, _ = fmt.Scan(&arr[i])
    }
    return arr
}

// k: length of array arrivalTimes
// T: O(k)
// S: O(1) extra space
func angryProfessor(arrivalTimes []int, threshold int) bool {
    onTime := 0
    for _, t := range arrivalTimes {
        if t <= 0 {
            onTime++
        }
    }
    return onTime < threshold
}
